src/error.ts

```diff
@@ -0,0 +1,95 @@
+import { Reply, type ReplyFrom, Status } from './reply'
+import { type Logger, recordException } from './logger'
+import type { ZodType } from 'zod'
+import type { Data } from './data'
+import type { MaybePromise } from './types'
+import type { ExtensionsFrom, MiddlewareContext, MiddlewareList } from './middleware'
+import type { ServerOptions } from './server'
+import { trace } from '@opentelemetry/api'
+import { name, version } from '../package.json'
+
+
+export type ErrorHandler<
+  Requires extends MiddlewareList,
+  Replies extends Partial<Record<Status, ZodType>> | undefined = undefined
+> = (input: {
+  event: Data & { reply: ReplyFrom<Replies>; error: Error } & MiddlewareContext & Partial<ExtensionsFrom<Requires>>
+  env: Environment
+  logger: Logger
+  error: Error
+}) => MaybePromise<Reply>
+
+export class CometErrorHandler {
+  private static readonly tracer = trace.getTracer(name, version)
+  static async handle(input: Omit<Parameters<ErrorHandler<any>>[0], 'error'>, error: unknown, options: ServerOptions<any, any, any>) {
+    return CometErrorHandler.tracer.startActiveSpan('comet error handler', async span => {
+      try {
+        const cometError = error instanceof Error ? error : new CometError(ErrorType.Internal, error)
+
+        if (options.errorHandler) {
+          span.setAttribute('handler', 'custom')
+
+          const reply = await options.errorHandler({ ...input, error: cometError })
+
+          return await Reply.toResponse(reply, options)
+        }
+
+        span.setAttribute('handler', 'internal')
+
+
+        return new CometErrorHandler({ ...input, error: cometError }).handleError()
+
+      } catch (error) {
+        recordException('[Comet] Failed to handle error.')
+        recordException(error)
+        span.setAttribute('handler', 'backup')
+
+        return new Response('An internal error has occurred.', { status: 500 })
+      }
+    })
+  }
+
+  constructor(private readonly input: Parameters<ErrorHandler<any>>[0]) {}
+
+  private async handleError(): Promise<Response> {
+    // record exception
+    this.input.logger.error(this.input.error)
+
+    // handle CometError
+    if (this.input.error instanceof CometError) {
+      switch (this.input.error.type) {
+        case ErrorType.MethodNotAllowed:
+          return new Response('Method not allowed.', { status: 405 })
+        case ErrorType.NotFound:
+          return new Response('Not found.', { status: 404 })
```

Comment:
Should we not return json here as well?

Reply:
@DaniFoldi

Reply:
Good idea, I've switched the default error handler to all json. The original idea was that it would be overridden anyway, but to match the old errors we can keep the default json.